### PR TITLE
add package qca

### PR DESCRIPTION
**Pull request by tehnick**

### CODE REVIEW

**Comment by tehnick** [comment]
Rebased onto current master.

Any comments?

**Comment by pavelvat** [comment]
>Any comments?

A few obvious notes:
1)replace
```
mkdir '$(1).build'
cd    '$(1).build' && $(TARGET)-cmake '$(1)' \
```
to
```
cd '$(BUILD_DIR)' && $(TARGET)-cmake '$(SOURCE_DIR)' \
```
2)replace
```
$(MAKE) -C '$(1).build' -j '$(JOBS)'
```
to
```
$(MAKE) -C '$(BUILD_DIR)' -j $(JOBS)
```
3)Maybe it
```
-DBUILD_STATIC_LIBS=$(CMAKE_STATIC_BOOL) \
-DBUILD_SHARED_LIBS=$(CMAKE_SHARED_BOOL)
```
is not needed since https://github.com/mxe/mxe/pull/1621
4)Section ```define $(PKG)_UPDATE``` must use ```MXE_GET_GITHUB_TAGS``` (some fixes in https://github.com/mxe/mxe/pull/1712)

**Comment by tehnick** [comment]
1)-3) Ok. Done.
4) It doesn't looks ready.

**Comment by starius** [comment]
It is installing the man page. Please apply the following patch to fix this:

```diff
diff --git a/src/qca.mk b/src/qca.mk
index 0216f16..893b974 100644
--- a/src/qca.mk
+++ b/src/qca.mk
@@ -22,7 +22,8 @@ define $(PKG)_BUILD
         -DBUILD_TESTS=OFF \
         -DBUILD_TOOLS=OFF \
         -DUSE_RELATIVE_PATHS=OFF \
-        -DBUILD_PLUGINS="auto"
+        -DBUILD_PLUGINS="auto" \
+        -DQCA_MAN_INSTALL_DIR="$(BUILD_DIR)/null"
     $(MAKE) -C '$(BUILD_DIR)' -j '$(JOBS)'
     $(MAKE) -C '$(BUILD_DIR)' -j 1 install
 endef
```

**Comment by tonytheodore** [comment]
> 4) It doesn't looks ready.

After #1712, you can replace `$(PKG)_FILE, $(PKG)_SUBDIR, $(PKG)_URL, and $(PKG)_UPDATE` with:
```
$(PKG)_GH_CONF  := KDE/qca, v
```

**Comment by starius** [comment]
Please add a test.

**Comment by tehnick** [comment]
> After #1712, you can replace $(PKG)_FILE, $(PKG)_SUBDIR, $(PKG)_URL, and $(PKG)_UPDATE with:
> ```
> $(PKG)_GH_CONF  := KDE/qca, v
> ```

```
boris@Tehnick:~/Opt/mxe__gcc5$ LC_ALL=C make qca
[download]  qca
[build]     qca                    i686-w64-mingw32.static

Failed to build package qca for target i686-w64-mingw32.static!
------------------------------------------------------------
loading initial cache file /home/boris/Opt/mxe__gcc5/usr/share/cmake/modules/TryRunResults.cmake
CMake Error: The source directory "/home/boris/Opt/mxe__gcc5/tmp-qca-i686-w64-mingw32.static/qca-v2.1.3" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
Makefile:632: ошибка выполнения рецепта для цели «build-only-qca_i686-w64-mingw32.static»
make[1]: *** [build-only-qca_i686-w64-mingw32.static] Ошибка 1
make[1]: выход из каталога «/home/boris/Opt/mxe__gcc5»
real    0m9,618s
user    0m9,348s
sys     0m0,096s
------------------------------------------------------------
[log]      /home/boris/Opt/mxe__gcc5/log/qca_i686-w64-mingw32.static

Makefile:632: recipe for target '/home/boris/Opt/mxe__gcc5/usr/i686-w64-mingw32.static/installed/qca' failed
make: *** [/home/boris/Opt/mxe__gcc5/usr/i686-w64-mingw32.static/installed/qca] Error 1
```

**Comment by tehnick** [comment]
I see tarball:
```
boris@Tehnick:~/Opt/mxe__gcc5$ ls pkg/*qca*
pkg/qca-v2.1.3.tar.gz
```
But after unpacking all sources are ing `qca-2.1.3`  subdirectory...

**Comment by tonytheodore** [comment]
#1729 fixes the `v` prefix issue.

I've done a PR https://github.com/sibuserv/mxe/pull/3 to build a simple test program, it fails at the moment but these are useful for future users as an example of how to link the package. It doesn't have to be based on `pkg-config`, it can be `cmake` or `qmake` - just some working example of linkage and flags.

**Comment by tehnick** [comment]
I have added test which is built in a three different ways.
Generated binary files (test-qca-*) are different but work fine.

**Comment by tehnick** [comment]
I forgot to mention that linking with static library currently looks impossible.
Probably static targets of qca package should be suppressed.

File: src/qca.mk [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    $(MAKE) -C '$(BUILD_DIR)' -j 1 install
+
+    # build test as qmake project
+    mkdir '$(BUILD_DIR)/test-qca'
+    cd '$(BUILD_DIR)/test-qca' && '$(PREFIX)/$(TARGET)/qt5/bin/qmake' '$(PWD)/src/qca-test.pro'
+    $(MAKE) -C '$(BUILD_DIR)/test-qca' -j $(JOBS) $(BUILD_TYPE)
+    $(INSTALL) -m755 '$(BUILD_DIR)/test-qca/$(BUILD_TYPE)/test-qca-qmake.exe' '$(PREFIX)/$(TARGET)/bin/'
+
+    # build test as cmake project
+    mkdir '$(1).test-cmake'
```
Comment (starius) [inline src/qca.mk]: IMHO, `mkdir '$(BUILD_DIR)/test-cmake'`

File: src/qca.mk [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    # build test manually
+    '$(TARGET)-g++' \
+        -W -Wall -Werror -std=gnu++11 \
+        '$(PWD)/src/qca-test.cpp' \
+        -o '$(PREFIX)/$(TARGET)/bin/test-$(PKG)-pkgconfig.exe' \
+        $(if $(BUILD_STATIC), -L'$(PREFIX)/$(TARGET)/qt5/plugins/crypto' -lqca-ossl) \
+        `'$(TARGET)-pkg-config' qca2-qt5 --cflags --libs`
+endef
+
```
Comment (starius) [inline src/qca.mk]: Extra empty line

File: src/qca-test.cmake [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+include_directories (${Qt5Core_INCLUDE_DIRS})
+target_link_libraries(${TGT} Qt5::Core qca-qt5)
+
+# Statically link QCA plugins when necessary
+if(BUILD_STATIC_LIBS)
+    target_link_libraries(${TGT} qca-ossl)
+endif(BUILD_STATIC_LIBS)
+
+install(TARGETS ${TGT} DESTINATION bin)
+
```
Comment (starius) [inline src/qca-test.cmake]: Extra empty line.

**Comment by starius** [comment]
> I forgot to mention that linking with static library currently looks impossible.
> Probably static targets of qca package should be suppressed.

What is the error?

**Comment by starius** [comment]
>> I forgot to mention that linking with static library currently looks impossible.
>> Probably static targets of qca package should be suppressed.

> What is the error?

https://gist.github.com/643efae7af89684954cf521df8d1f124

It looks like a missing library in linker options.

**Comment by starius** [comment]
Shared targets tested.

**Comment by tehnick** [comment]
> It looks like a missing library in linker options.

No. -lqca-qt5 is present in linker options and placed on the appropriate place.

File: src/qca-test.cmake (hunk 3 above)
Comment (tehnick) [inline src/qca-test.cmake]: Just my code style...

**Comment by tehnick** [comment]
Updated.

**Comment by tonytheodore** [comment]
Seems to have something to do with name mangling and not setting the correct [export macros](https://github.com/KDE/qca/blob/master/include/QtCrypto/qca_export.h#L40)
for [static](https://github.com/KDE/qca/blob/master/CMakeLists.txt#L92) and [shared](https://github.com/KDE/qca/blob/master/src/CMakeLists.txt#L170). Probably need upstream input to get the static builds working.

This looks good to merge, thanks!

**Pull request merged**